### ISSUE TITLE
Ideas for guiding devs to build performant apps and avoid common mistakes.

### ISSUE THREAD

**Issue by InfamousVague** [DioxusLabs/dioxus#623]
Was mulling this over the past few days; the biggest struggle we're having developing on top of Rust and Dioxus itself (not a fault of either of course) is bringing traditional web folks into the world of Rust; I was curious, would it be possible to convert things like `componentDidMount` or whatever the equivalent is now days into a scoped block that executes when the component is rendered into view, and "supporting" other features to get basic functionality. I suspect this will largely create confusion and anti-patterns carried over from JS but wanted to get the feedback of people more intelligent than myself before asking some poor soul on our team to go down trying to modify ASTs and whatnot. 

I think the hardest thing to grasp has been (for me at least, a total noob to building anything real in rust) is the automatic execution of the code before the rsx macro on every re-render, once it's understood a lot of performance mistakes make themselves apparent and you start to understand why you're running at 200% cpu 😅. 

**TL;DR is there way we could help to onboard new devs (to the world of rust and dioxus) to help guide them to build performant applications using Dioxus and avoiding common mistakes that can cause larger issues and needless re-rendering of components? If so, is there any way we can help contribute to this effort?**

Thank you in advanced for reading!

**Comment by Demonthos** [comment]
# Node Refs
For Node refs and onmounted I opened a new issue to discuss ideas: https://github.com/DioxusLabs/dioxus/issues/631
We need some sort of solution here but it is probably best to find an api that works well both for rust and many different platforms before we implement anything. Different platforms handle nodes very differently so a small subset will probably be added first and then expanded. There are also serialization boundaries on some platforms (webview and liveview) that might make synchronous query's preform poorly.

# Performance

The mental model of react/dioxus is very different from traditional programming. Docs is probably the easiest thing we can improve here, but there are also some areas we can improve tooling.

## Docs
We could add more examples to the docs in the best practices section. It might be useful to see how react handles onboarding new devs with the [new beta docs](https://beta.reactjs.org/learn/state-a-components-memory).


It's also worth noting that the master version of dioxus performs very differently from the ```0.2.4``` version and react. The previous version diffed every node every time; the master version only diffs dynamic nodes. This makes performance a lot better for most apps as we need to do less diffing, but for apps with similar branches it will preform worse. We can provide some diffing in the new version, but the more diffing we do the less of a benefit template will be.

For example in this code if x changes from true to false the div and text node will change in the dom in the master version. In the 0.2.4 version only the text would change.
```rust
if x {
  render!{
    div{ "0" }
  }
}else {
  render!{
    div{ "1" }
  }
}
```

What users should do instead is:
```rust
let y = if x {0} else{1};
render!{
  div{
    "{x}"
  }
}
```

The Docs should probably emphasize this for the next version.

Here are the relevant sections the guide: 
Performance of updating state during rendering: https://dioxuslabs.com/nightly/guide/best_practices/antipatterns.html#avoid-updating-state-during-render
Hooks causing re-renders: https://dioxuslabs.com/nightly/guide/interactivity/hooks.html#hooks-and-component-state

## Tooling
There are some tooling areas we can improve to make it more clear what is causing re-renders:
- Adding a method that gives a reason for re-rendering https://github.com/DioxusLabs/dioxus/issues/475
- A browser extension that visually shows how state flows between components. This would be more fine-grained and could provide a timeline of events (this event caused this specific state to update which re-rendered this component and caused this other state to update)
